mula/scheduler/connectors/listeners/listeners.py

```diff
@@ -56,32 +67,49 @@ def __init__(self, dsn: str):
                 connect to.
         """
         super().__init__()
-        self.dsn = dsn
 
-        self.connection = pika.BlockingConnection(pika.URLParameters(self.dsn))
-        self.channel = self.connection.channel()
+        self.dsn = dsn
+        self.executor: futures.ThreadPoolExecutor = futures.ThreadPoolExecutor(max_workers=10)
 
-    def dispatch(self, body: bytes) -> None:
+    def dispatch(self, channel, delivery_tag, body: bytes) -> None:
         """Dispatch a message without a return value"""
         raise NotImplementedError
 
     def basic_consume(self, queue: str, durable: bool, prefetch_count: int) -> None:
-        try:
-            self.channel.queue_declare(queue=queue, durable=durable)
-        except pika.exceptions.ChannelClosedByBroker as exc:
-            if "inequivalent arg 'durable'" in exc.reply_text:
-                # Queue changed from non-durable to durable. Given that
-                # previously they weren't durable and contents would also be
-                # lost if RabbitMQ restarted, we will just delete the queue and
-                # recreate it to provide for a smooth upgrade.
-                self.channel = self.connection.channel()
-                self.channel.queue_delete(queue=queue)
+        while True:
+            self.connection = pika.BlockingConnection(pika.URLParameters(self.dsn))
+            self.channel = self.connection.channel()
+
+            try:
                 self.channel.queue_declare(queue=queue, durable=durable)
-            else:
-                raise
-        self.channel.basic_qos(prefetch_count=prefetch_count)
-        self.channel.basic_consume(queue, on_message_callback=self.callback)
-        self.channel.start_consuming()
+            except pika.exceptions.ChannelClosedByBroker as exc:
+                if "inequivalent arg 'durable'" in exc.reply_text:
+                    # Queue changed from non-durable to durable. Given that
+                    # previously they weren't durable and contents would also be
+                    # lost if RabbitMQ restarted, we will just delete the queue and
+                    # recreate it to provide for a smooth upgrade.
+                    self.channel = self.connection.channel()
+                    self.channel.queue_delete(queue=queue)
+                    self.channel.queue_declare(queue=queue, durable=durable)
+                else:
+                    raise
+
+            try:
+                self.channel.basic_qos(prefetch_count=prefetch_count)
+                self.channel.basic_consume(queue, on_message_callback=self.callback)
+                self.channel.start_consuming()
+            except pika.exceptions.ConnectionClosedByBroker:
+                # Recover from server-initiated connection closure, this also
```

Comment:
Hm, I'd say graceful shutdown of nodes could be due to maintenance as well. In this case it would be nice to have the scheduler try to reconnect a few times at least and recover when possible? However, we still have the following issue: there is no error handling for `self.reconnect()` here, which can throw a `StreamLostError`. This can get a bit messy though since we have to add retries/backoffs for this (or hard-code retrying twice as I implemented with a double `try`-`except` in `octopoes.events.manager` before..  There haven't been a lot if issues with it though)

Reply:
I've implemented the [retry](https://github.com/invl/retry) which seems to be a nice way to implement a retry protocol on certain exceptions.